src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AdminController.cs

```diff
@@ -127,12 +153,15 @@ public class AdminController : Controller
                 ViewData["ReturnUrl"] = returnUrl;
                 return View(model);
             }
+            
+            await TryUpdateModelAsync(application);
+            if (model.UpdateClientSecret && model.Type == ClientType.Confidential)
+                await _applicationManager.SetClientSecretAsync(application, model.ClientSecret, HttpContext.RequestAborted);
 
-            var application = await _applicationManager.FindByIdAsync(model.Id, HttpContext.RequestAborted);
-            if (application == null)
-                return NotFound();
+            application.NormalizedRoleNames = new List<string>();
+            if (application.Type == ClientType.Confidential && application.AllowClientCredentialsFlow)
+                application.NormalizedRoleNames = model.RoleEntries.Where(r => r.Selected).Select(r => r.NormalizedName).ToList();
 
-            await TryUpdateModelAsync(application);
             await _applicationManager.UpdateAsync(application, HttpContext.RequestAborted);
```

Comment:
I have tried to avoid problem of UpdateAsync method that throws an exception if you attempt to modify a confidential app because it doesn't have a client secret calling to await _applicationManager.SetClientSecretAsync() at line 159, but this also returns an exception. So we need a fix on Openiddict.
I left the code with this attempt because no matter if it is there or not the exception will always happen.